### PR TITLE
test: remaining incentive module unit tests

### PR DESCRIPTION
**Pull request by toteki**
closes: #2005 
progress: #1990

Note: What remains in #1990 is the leverage-side tests (basically `hooks_test.go` but run from the leverage module during its integration tests) and some edge cases.

### CODE REVIEW

File: x/incentive/keeper/mock_test.go [hunk 1]
```diff
@@ -148,7 +149,7 @@ func (m *mockLeverageKeeper) GetCollateral(_ sdk.Context, addr sdk.AccAddress, d
 func (m *mockLeverageKeeper) DonateCollateral(ctx sdk.Context, addr sdk.AccAddress, uToken sdk.Coin) error {
 	newCollateral := m.GetCollateral(ctx, addr, uToken.Denom).Sub(uToken).Amount.Int64()
 	m.setCollateral(addr, uToken.Denom, newCollateral)
-	m.donatedCollateral = m.donatedCollateral.Add(uToken)
+	*m.donatedCollateral = m.donatedCollateral.Add(uToken)
```
Comment (toteki) [inline x/incentive/keeper/mock_test.go:152]: Fixes a bug in the mock keeper where `k.msgSrv.leveragekeeper` and `k.lk` were storing independent `donatedCollateral`, so one would update and the other not depending on how `DonateCollateral` was called.

File: x/incentive/keeper/msg_server.go [hunk 2]
```diff
@@ -131,7 +131,7 @@ func (s msgServer) EmergencyUnbond(
 		return nil, err
 	}
 
-	maxEmergencyUnbond := k.restrictedCollateral(ctx, addr, msg.UToken.Denom)
```
Comment (toteki) [inline x/incentive/keeper/msg_server.go:134]: simplification (these were always the same string)

File: x/incentive/keeper/unbonding.go [hunk 3]
```diff
@@ -98,11 +98,11 @@ func (k Keeper) reduceBondTo(ctx sdk.Context, addr sdk.AccAddress, newCollateral
 	// if we have not returned yet, the only some in-progress unbondings will be
 	// instantly unbonded.
 	amountToUnbond := bonded.Amount.Add(unbonding.Amount).Sub(newCollateral.Amount)
-	for _, u := range unbondings {
+	for i, u := range unbondings {
 		// for ongoing unbondings, starting with the oldest
 		specificReduction := sdk.MinInt(amountToUnbond, u.UToken.Amount)
 		// reduce the in-progress unbonding amount, and the remaining instant unbond
-		u.UToken.Amount = u.UToken.Amount.Sub(specificReduction)
+		unbondings[i].UToken.Amount = u.UToken.Amount.Sub(specificReduction)
```
Comment (toteki) [inline x/incentive/keeper/unbonding.go:105]: bug fixed (unbondings were not being set)

File: x/incentive/genesis.go [hunk 4]
```diff
@@ -58,54 +59,98 @@ func (gs GenesisState) Validate() error {
...
 
 	return nil
 }
 
+// noDuplicateString checks to see if a string is already present in a map
+// and then adds it to the map. If it was already present, an error is returned.
+// used to check all uniqueness requirements in genesis state.
+func noDuplicateString(m map[string]bool, s, errMsg string) error {
+	if _, found := m[s]; found {
+		return fmt.Errorf("duplicate %s: %s", errMsg, s)
```
Comment (robert-zaremba) [inline x/incentive/genesis.go:148]:
```suggestion
		return fmt.Errorf("duplicate %s: %s", s, errMsg)
```

File: x/incentive/genesis_test.go [hunk 5]
```diff
@@ -4,9 +4,277 @@ import (
...
 
 	"gotest.tools/v3/assert"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	"github.com/umee-network/umee/v4/util/coin"
+	leveragetypes "github.com/umee-network/umee/v4/x/leverage/types"
 )
 
 func TestValidateGenesis(t *testing.T) {
```
Comment (robert-zaremba) [inline x/incentive/genesis_test.go:14]: Add `t.Parallel()` to all unit tests (at the very beginning).

**Review by robert-zaremba — COMMENTED** [review]
start of the review

File: x/incentive/genesis.go (hunk 4 above)
Comment (robert-zaremba) [reply x/incentive/genesis.go:148]: ah, ok, when reading my impression was that the details are in `s`. Let's keep as is.

File: x/incentive/genesis_test.go [hunk 6]
```diff
@@ -4,9 +4,277 @@ import (
...
+	validBond := NewBond(validAddr, coin.New("u/uumee", 1))
+	assert.NilError(t, validBond.Validate())
+
+	invalidBond := validBond
+	invalidBond.Account = ""
+	assert.ErrorContains(t, invalidBond.Validate(), "empty address string is not allowed")
+
+	invalidBond = validBond
+	invalidBond.UToken = sdk.Coin{}
+	assert.ErrorContains(t, invalidBond.Validate(), "invalid denom")
```
[the remark this answers is not in the document]
Comment (toteki) [reply x/incentive/genesis_test.go:188]: sdk.Coin.Validate iis called first, so `denom` is the error we get

File: x/incentive/genesis_test.go [hunk 7]
```diff
@@ -4,9 +4,277 @@ import (
...
+	proposedRemainingRewards.RemainingRewards = coin.New("uumee", 1)
+	assert.ErrorIs(t, proposedRemainingRewards.ValidateProposed(), ErrNonzeroRemainingRewards, "proposed remaining rewards")
+
+	invalidProposed := validProposed
+	invalidProposed.StartTime = 0
+	assert.ErrorIs(t, invalidProposed.ValidateProposed(), ErrInvalidProgramStart, "proposed invalid program")
+
+	proposedFunded := validProposed
+	proposedFunded.Funded = true
+	assert.ErrorIs(t, proposedFunded.ValidateProposed(), ErrProposedFundedProgram, "proposed funded program")
```
[the remark this answers is not in the document]
Comment (toteki) [reply x/incentive/genesis_test.go:169]: A program, when proposed, has not yet been funded by either community fund or manual method (so its `Funded` must be false at time of proposal)

File: x/incentive/genesis_test.go [hunk 8]
```diff
@@ -4,9 +4,277 @@ import (
 	"testing"
 
 	"gotest.tools/v3/assert"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	"github.com/umee-network/umee/v4/util/coin"
```
[the remark this answers is not in the document]
Comment (toteki) [reply x/incentive/genesis_test.go]: done (merged main and caught a few stragglers)

File: x/incentive/genesis_test.go [hunk 9]
```diff
@@ -4,9 +4,277 @@ import (
...
+
+func TestValidateIncentiveProgram(t *testing.T) {
+	validProgram := NewIncentiveProgram(1, 1, 1, "u/uumee", coin.New("uumee", 1), coin.Zero("uumee"), false)
+	assert.NilError(t, validProgram.Validate())
+
+	invalidUToken := validProgram
+	invalidUToken.UToken = ""
+	assert.ErrorContains(t, invalidUToken.Validate(), "invalid denom")
+
+	invalidUToken.UToken = "uumee"
```
[the remark this answers is not in the document]
Comment (toteki) [reply x/incentive/genesis_test.go]: fixed instances that I found
